bundles/org.openhab.binding.nobohub/src/main/java/org/openhab/binding/nobohub/internal/ComponentHandler.java

```diff
@@ -0,0 +1,160 @@
+            updateProperty(PROPERTY_ZONE, zoneName);
+        }
+
+        String tempForZoneName = getZoneName(component.getTemperatureSensorForZoneId());
+        if (tempForZoneName != null) {
+            updateProperty(PROPERTY_TEMPERATURE_SENSOR_FOR_ZONE, tempForZoneName);
+        }
+    }
+
+    private @Nullable String getZoneName(int zoneId) {
+        Bridge noboHub = getBridge();
+        if (null != noboHub) {
+            NoboHubBridgeHandler hubHandler = (NoboHubBridgeHandler) noboHub.getHandler();
+            if (hubHandler != null) {
+                Zone zone = hubHandler.getZone(zoneId);
+                if (null != zone) {
+                    return zone.getName();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public void initialize() {
+        String serialNumberString = getConfigAs(ComponentConfiguration.class).serialNumber;
+        if (serialNumberString != null && !serialNumberString.isEmpty()) {
+            SerialNumber sn = new SerialNumber(serialNumberString);
+            if (!sn.isWellFormed()) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "@text/message.component.illegal.serial [\"" + serialNumber + "\"]");
+            } else {
+                this.serialNumber = sn;
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "@text/message.missing.serial");
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            logger.debug("Refreshing channel {}", channelUID);
+            if (null != serialNumber) {
+                Component component = getComponent();
+                if (null == component) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.GONE,
+                            messages.getText("message.component.notfound", serialNumber, channelUID));
+                } else {
+                    onUpdate(component);
+                }
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.GONE,
+                        "@text/message.component.missing.id [\"" + channelUID + "\"]");
+            }
+
+            return;
+        }
+
+        logger.debug("This component is a read-only device and cannot handle commands.");
+    }
+
+    public @Nullable SerialNumber getSerialNumber() {
+        return serialNumber;
+    }
+
+    private @Nullable Component getComponent() {
+        Bridge noboHub = getBridge();
+        if (null != noboHub) {
+            NoboHubBridgeHandler hubHandler = (NoboHubBridgeHandler) noboHub.getHandler();
+            if (null != serialNumber && null != hubHandler) {
+                SerialNumber sn = Helpers.castToNonNull(serialNumber, Thing.PROPERTY_SERIAL_NUMBER);
```

Comment:
Why is this cast needed?

Reply:
Ah, I see. Please see my proposal below. This applies to all places where you are using this `castToNonNull` construct. The reason null annotation complains is because another thread could have set the instance variable to null between your null check and usage. The most common way to mitigate this is to copy into a local variable. Your `castToNonNull` only hides this problem, but doesn't mitigate it - thus you still risk having `IllegalArgumentException` thrown runtime (instead of `NullPointerException`).